### PR TITLE
fix: add concurrency control for catalog manager

### LINKED ISSUE #617: Bug: Creating the table in concurrent may cause the table id wrong in catalog manager
### What type of bug is this?

Data corruption

### What subsystems are affected?

Datanode

### What happened?

In Prometheus or InfuxDB cases, GreptimeDB will create the metric tables automatically by requests in concurrent, and the catalog manager will allocate a table id for every request:

```
2022-11-22T07:41:43.466429Z  INFO frontend::instance: Table greptime.prometheus.prometheus_http_response_size_bytes_bucket does not exist, try create table
2022-11-22T07:41:43.466471Z  INFO frontend::instance: Try to create table: prometheus_http_response_size_bytes_bucket automatically with request: CreateExpr { catalog_name: Some("greptime"), schema_name: Some("prometheus"), table_name: "prometheus_http_response_size_bytes_bucket", desc: Some("Created on insertion"), column_defs: [ColumnDef { name: "greptime_timestamp", datatype: Timestamp, is_nullable: false, default_constraint: None }, ColumnDef { name: "greptime_value", datatype: Float64, is_nullable: true, default_constraint: None }, ColumnDef { name: "handler", datatype: String, is_nullable: true, default_constraint: None }, ColumnDef { name: "instance", datatype: String, is_nullable: true, default_constraint: None }, ColumnDef { name: "job", datatype: String, is_nullable: true, default_constraint: None }, ColumnDef { name: "le", datatype: String, is_nullable: true, default_constraint: None }], time_index: "greptime_timestamp", primary_keys: ["handler", "instance", "job", "le"], create_if_not_exists: true, table_options: {}, table_id: None, region_ids: [0] }
2022-11-22T07:41:43.466565Z  INFO frontend::instance: Table greptime.prometheus.prometheus_http_response_size_bytes_bucket does not exist, try create table
2022-11-22T07:41:43.466610Z  INFO frontend::instance: Try to create table: prometheus_http_response_size_bytes_bucket automatically with request: CreateExpr { catalog_name: Some("greptime"), schema_name: Some("prometheus"), table_name: "prometheus_http_response_size_bytes_bucket", desc: Some("Created on insertion"), column_defs: [ColumnDef { name: "greptime_timestamp", datatype: Timestamp, is_nullable: false, default_constraint: None }, ColumnDef { name: "greptime_value", datatype: Float64, is_nullable: true, default_constraint: None }, ColumnDef { name: "handler", datatype: String, is_nullable: true, default_constraint: None }, ColumnDef { name: "instance", datatype: String, is_nullable: true, default_constraint: None }, ColumnDef { name: "job", datatype: String, is_nullable: true, default_constraint: None }, ColumnDef { name: "le", datatype: String, is_nullable: true, default_constraint: None }], time_index: "greptime_timestamp", primary_keys: ["handler", "instance", "job", "le"], create_if_not_exists: true, table_options: {}, table_id: None, region_ids: [0] }
2022-11-22T07:41:43.466709Z  INFO frontend::instance: Table greptime.prometheus.prometheus_http_response_size_bytes_bucket does not exist, try create table
2022-11-22T07:41:43.466750Z  INFO frontend::instance: Try to create table: prometheus_http_response_size_bytes_bucket automatically with request: CreateExpr { catalog_name: Some("greptime"), schema_name: Some("prometheus"), table_name: "prometheus_http_response_size_bytes_bucket", desc: Some("Created on insertion"), column_defs: [ColumnDef { name: "greptime_timestamp", datatype: Timestamp, is_nullable: false, default_constraint: None }, ColumnDef { name: "greptime_value", datatype: Float64, is_nullable: true, default_constraint: None }, ColumnDef { name: "handler", datatype: String, is_nullable: true, default_constraint: None }, ColumnDef { name: "instance", datatype: String, is_nullable: true, default_constraint: None }, ColumnDef { name: "job", datatype: String, is_nullable: true, default_constraint: None }, ColumnDef { name: "le", datatype: String, is_nullable: true, default_constraint: None }], time_index: "greptime_timestamp", primary_keys: ["handler", "instance", "job", "le"], create_if_not_exists: true, table_options: {}, table_id: None, region_ids: [0] }
2022-11-22T07:41:43.466845Z  INFO frontend::instance: Table greptime.prometheus.prometheus_http_response_size_bytes_bucket does not exist, try create table
2022-11-22T07:41:43.466884Z  INFO frontend::instance: Try to create table: prometheus_http_response_size_bytes_bucket automatically with request: CreateExpr { catalog_name: Some("greptime"), schema_name: Some("prometheus"), table_name: "prometheus_http_response_size_bytes_bucket", desc: Some("Created on insertion"), column_defs: [ColumnDef { name: "greptime_timestamp", datatype: Timestamp, is_nullable: false, default_constraint: None }, ColumnDef { name: "greptime_value", datatype: Float64, is_nullable: true, default_constraint: None }, ColumnDef { name: "handler", datatype: String, is_nullable: true, default_constraint: None }, ColumnDef { name: "instance", datatype: String, is_nullable: true, default_constraint: None }, ColumnDef { name: "job", datatype: String, is_nullable: true, default_constraint: None }, ColumnDef { name: "le", datatype: String, is_nullable: true, default_constraint: None }], time_index: "greptime_timestamp", primary_keys: ["handler", "instance", "job", "le"], create_if_not_exists: true, table_options: {}, table_id: None, region_ids: [0] }
2022-11-22T07:41:43.466984Z  INFO frontend::instance: Table greptime.prometheus.prometheus_http_response_size_bytes_bucket does not exist, try create table
2022-11-22T07:41:43.467029Z  INFO frontend::instance: Try to create table: prometheus_http_response_size_bytes_bucket automatically with request: CreateExpr { catalog_name: Some("greptime"), schema_name: Some("prometheus"), table_name: "prometheus_http_response_size_bytes_bucket", desc: Some("Created on insertion"), column_defs: [ColumnDef { name: "greptime_timestamp", datatype: Timestamp, is_nullable: false, default_constraint: None }, ColumnDef { name: "greptime_value", datatype: Float64, is_nullable: true, default_constraint: None }, ColumnDef { name: "handler", datatype: String, is_nullable: true, default_constraint: None }, ColumnDef { name: "instance", datatype: String, is_nullable: true, default_constraint: None }, ColumnDef { name: "job", datatype: String, is_nullable: true, default_constraint: None }, ColumnDef { name: "le", datatype: String, is_nullable: true, default_constraint: None }], time_index: "greptime_timestamp", primary_keys: ["handler", "instance", "job", "le"], create_if_not_exists: true, table_options: {}, table_id: None, region_ids: [0] }
``` 

```
2022-11-22T07:32:56.470138Z  INFO datanode::server::grpc::ddl: Creating table Some("greptime").Some("prometheus")."prometheus_http_requests_total" with table id from catalog manager: 1454
2022-11-22T07:32:56.470193Z  INFO datanode::server::grpc::ddl: Creating table Some("greptime").Some("prometheus")."prometheus_http_response_size_bytes_bucket" with table id from catalog manager: 1455
2022-11-22T07:32:56.470261Z  INFO datanode::server::grpc::ddl: Creating table Some("greptime").Some("prometheus")."prometheus_http_response_size_bytes_bucket" with table id from catalog manager: 1457
2022-11-22T07:32:56.470292Z  INFO datanode::server::grpc::ddl: Creating table Some("greptime").Some("prometheus")."prometheus_http_response_size_bytes_bucket" with table id from catalog manager: 1458
2022-11-22T07:32:56.470266Z  INFO datanode::server::grpc::ddl: Creating table Some("greptime").Some("prometheus")."prometheus_http_request_duration_seconds_count" with table id from catalog manager: 1456
2022-11-22T07:32:56.470327Z  INFO datanode::server::grpc::ddl: Creating table Some("greptime").Some("prometheus")."prometheus_http_response_size_bytes_bucket" with table id from catalog manager: 1459
2022-11-22T07:32:56.470351Z  INFO datanode::server::grpc::ddl: Creating table Some("greptime").Some("prometheus")."prometheus_http_response_size_bytes_bucket" with table id from catalog manager: 1460
2022-11-22T07:32:56.470380Z  INFO datanode::server::grpc::ddl: Creating table Some("greptime").Some("prometheus")."prometheus_http_response_size_bytes_bucket" with table id from catalog manager: 1461
2022-11-22T07:32:56.470383Z  INFO datanode::server::grpc::ddl: Creating table Some("greptime").Some("prometheus")."prometheus_http_response_size_bytes_bucket" with table id from catalog manager: 1462
2022-11-22T07:32:56.470394Z  INFO datanode::server::grpc::ddl: Creating table Some("greptime").Some("prometheus")."prometheus_http_response_size_bytes_bucket" with table id from catalog manager: 1463
2022-11-22T07:32:56.470484Z  INFO datanode::server::grpc::ddl: Creating table Some("greptime").Some("prometheus")."prometheus_http_response_size_bytes_bucket" with table id from catalog manager: 1464
```

Then the table engine will create the table one and only one successfully, but return the created table reference when `create_if_not_exists: true`. In such a case, the catalog manager will insert many duplicated table entries with different table IDs, and the last one succeed. 

When restarting the datanode, the catalog manager will try to open the table with the wrong table id and panic:

```
2022-11-22T07:41:45.243560Z  WARN opendal::services: service=fs operation=stat path=prometheus/prometheus_http_response_size_bytes_bucket/1464_0000000000/manifest/ -> errored: Custom { kind: NotFound, error: ObjectError { op: Stat, path: "prometheus/prometheus_http_response_size_bytes_bucket/1464_0000000000/manifest/", source: No such file or directory (os error 2) } }
2022-11-22T07:41:45.629782Z ERROR greptime: Fatal error occurs! err.msg=Failed to start datanode, source: Failed to create catalog list, source: Table not found while opening table, table info: greptime.prometheus.prometheus_http_response_size_bytes_bucket, id: 1464 err.code=Unexpected err.source=Failed to create catalog list, source: Table not found while opening table, table info: greptime.prometheus.prometheus_http_response_size_bytes_bucket, id: 1464 err.source.sources=[Table not found while opening table, table info: greptime.prometheus.prometheus_http_response_size_bytes_bucket, id: 1464] err.backtrace= 0 backtrace::backtrace::libunwind::trace::hcdfd8b20f85c976d
   /Users/dennis/.cargo/registry/src/github.com-1ecc6299db9ec823/backtrace-0.3.66/src/backtrace/libunwind.rs:93
   backtrace::backtrace::trace_unsynchronized::h92cf28ac583400a8
   /Users/dennis/.cargo/registry/src/github.com-1ecc6299db9ec823/backtrace-0.3.66/src/backtrace/mod.rs:66
 1 backtrace::backtrace::trace::h6b7004ab3cf91bc8
   /Users/dennis/.cargo/registry/src/github.com-1ecc6299db9ec823/backtrace-0.3.66/src/backtrace/mod.rs:53
 2 backtrace::capture::Backtrace::create::h3c74948d4e1cf141
   /Users/dennis/.cargo/registry/src/github.com-1ecc6299db9ec823/backtrace-0.3.66/src/capture.rs:176
 3 backtrace::capture::Backtrace::new::h7226af6750c0259f
   /Users/dennis/.cargo/registry/src/github.com-1ecc6299db9ec823/backtrace-0.3.66/src/capture.rs:140
 4 <snafu::backtrace_shim::Backtrace as snafu::GenerateImplicitData>::generate::h56a4c240894038fb
   /Users/dennis/.cargo/registry/src/github.com-1ecc6299db9ec823/snafu-0.7.1/src/backtrace_shim.rs:15
   <catalog::error::TableNotFoundSnafu<__T0> as snafu::IntoError<catalog::error::Error>>::into_error::h8225d6db6eb9f79b
   /Users/dennis/programming/rust/greptimedb/src/catalog/src/error.rs:24
 5 <core::option::Option<T> as snafu::OptionExt<T>>::with_context::hd5e149afdb023e82
   /Users/dennis/.cargo/registry/src/github.com-1ecc6299db9ec823/snafu-0.7.1/src/lib.rs:913
 6 catalog::local::manager::LocalCatalogManager::open_and_register_table::{{closure}}::h2907bce62c45e814
   /Users/dennis/programming/rust/greptimedb/src/catalog/src/local/manager.rs:247
 7 <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll::h089dcb678a0f3637
   /rustc/87588a2afd9ca903366f0deaf84d805f34469384/library/core/src/future/mod.rs:91
 8 catalog::local::manager::LocalCatalogManager::handle_system_catalog_entries::{{closure}}::h1034d60d66c6decb
   /Users/dennis/programming/rust/greptimedb/src/catalog/src/local/manager.rs:209
 9 <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll::hccb6ebfe744f3a1a
   /rustc/87588a2afd9ca903366f0deaf84d805f34469384/library/core/src/future/mod.rs:91
10 catalog::local::manager::LocalCatalogManager::init::{{closure}}::hf9b5c984b9d873f2
   /Users/dennis/programming/rust/greptimedb/src/catalog/src/local/manager.rs:88
11 <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll::h7f01b9605513b4f1
   /rustc/87588a2afd9ca903366f0deaf84d805f34469384/library/core/src/future/mod.rs:91
12 <catalog::local::manager::LocalCatalogManager as catalog::CatalogManager>::start::{{closure}}::hcd2185fa73ac7c2f
   /Users/dennis/programming/rust/greptimedb/src/catalog/src/local/manager.rs:309
13 <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll::hb2040732e99ab48c
   /rustc/87588a2afd9ca903366f0deaf84d805f34469384/library/core/src/future/mod.rs:91
14 <core::pin::Pin<P> as core::future::future::Future>::poll::hab1ed86d29c62555
   /rustc/87588a2afd9ca903366f0deaf84d805f34469384/library/core/src/future/future.rs:124
15 datanode::instance::Instance::start::{{closure}}::h0bf36126e450f32b
   /Users/dennis/programming/rust/greptimedb/src/datanode/src/instance.rs:154
16 <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll::h22422bd329d53c4d
   /rustc/87588a2afd9ca903366f0deaf84d805f34469384/library/core/src/future/mod.rs:91
17 datanode::datanode::Datanode::start::{{closure}}::hc75fc13c63a62838
   /Users/dennis/programming/rust/greptimedb/src/datanode/src/datanode.rs:89
18 <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll::hfecd6f54a0e62c2f
   /rustc/87588a2afd9ca903366f0deaf84d805f34469384/library/core/src/future/mod.rs:91
19 cmd::standalone::StartCommand::run::{{closure}}::{{closure}}::hd86e03db0f32aa43
   /Users/dennis/programming/rust/greptimedb/src/cmd/src/standalone.rs:160
20 <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll::hba77370dc3da740c
   /rustc/87588a2afd9ca903366f0deaf84d805f34469384/library/core/src/future/mod.rs:91
21 <tokio::future::maybe_done::MaybeDone<Fut> as core::future::future::Future>::poll::hc404e671e1baf1fd
   /Users/dennis/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.21.1/src/future/maybe_done.rs:68
22 cmd::standalone::StartCommand::run::{{closure}}::{{closure}}::h877d6f9d979e4960
   /Users/dennis/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.21.1/src/macros/try_join.rs:162
23 <tokio::future::poll_fn::PollFn<F> as core::future::future::Future>::poll::h1a9bb79bdf5535d6
   /Users/dennis/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.21.1/src/future/poll_fn.rs:38
24 cmd::standalone::StartCommand::run::{{closure}}::h25f08077d0d6fbf8
   /Users/dennis/programming/rust/greptimedb/src/cmd/src/standalone.rs:159
25 <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll::hcd52f4bdf4e3e411
   /rustc/87588a2afd9ca903366f0deaf84d805f34469384/library/core/src/future/mod.rs:91
26 cmd::standalone::SubCommand::run::{{closure}}::ha33bc4b23fce4afd
   /Users/dennis/programming/rust/greptimedb/src/cmd/src/standalone.rs:57
27 <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll::h451651784646db0d
   /rustc/87588a2afd9ca903366f0deaf84d805f34469384/library/core/src/future/mod.rs:91
28 cmd::standalone::Command::run::{{closure}}::ha270a189d27bf4cc
   /Users/dennis/programming/rust/greptimedb/src/cmd/src/standalone.rs:45
29 <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll::h9504e6e3c1d6dda8
   /rustc/87588a2afd9ca903366f0deaf84d805f34469384/library/core/src/future/mod.rs:91
30 greptime::SubCommand::run::{{closure}}::h2601c233c3671384
   /Users/dennis/programming/rust/greptimedb/src/cmd/src/bin/greptime.rs:57
```

I think the catalog manager should take care of this case to avoid duplicated table entires.

### What operating system did you use?

mac M1

### Relevant log output and stack trace

_No response_

### How can we reproduce the bug?

```bash
Create table concurrently and restart the data node.
```

### PR DESCRIPTION
**Pull request by v0y4g3r**
I hereby agree to the terms of the [GreptimeDB CLA](https://gist.github.com/xtang/6378857777706e568c1949c7578592cc)

## What's changed and what's your intention?

Implement concurrency control for catalog manager.
- In local catalog manager, check if registering table's id is the same with existing table, and return an error if not. The check is guarded by a mutex to prevent concurrent register.
- In remote catalog manager, use kvbackend's compare and set API to prevent registering tables with  same name but different id.

## Checklist

- [x]  I have written the necessary rustdoc comments.
- [x]  I have added the necessary unit tests and integration tests.

## Refer to a related PR or issue link (optional)
Fixes #617

### CODE REVIEW

File: src/catalog/src/local/manager.rs [hunk 1]
```diff
@@ -332,24 +334,36 @@ impl CatalogManager for LocalCatalogManager {
...
+                        table: format_full_table_name(
+                            catalog_name,
+                            schema_name,
+                            &request.table_name,
+                        ),
+                    }
+                    .fail();
+                }
+                // Try to register table with same table id, just ignore.
+                Ok(0)
```
Comment (MichaelScofield) [inline src/catalog/src/local/manager.rs:352]: Any particular thought we have to check whether table_id is the same? I think checking table_name is enough.

Comment (v0y4g3r) [reply src/catalog/src/local/manager.rs:352]: Duplicated register for same table name with the same table id is ok and catalog manager just ignores it.
But registering tables with same table name but different table id is considered abnormal, which indicates that table engine has created same table name with different table id, which should be avoided by table engine.

File: src/frontend/src/instance/distributed.rs [hunk 2]
```diff
@@ -236,17 +236,32 @@ impl DistInstance {
...
         let value = create_table_global_value(create_table, table_route)?
             .as_bytes()
             .context(error::CatalogEntrySerdeSnafu)?;
 
-        self.catalog_manager
+        if let Err(existing) = self
+            .catalog_manager
             .backend()
-            .set(key.to_string().as_bytes(), &value)
+            .compare_and_set(key.as_bytes(), &[], &value)
```
[the remark this answers is not in the document]
Comment (v0y4g3r) [reply src/frontend/src/instance/distributed.rs:249]: > No other errors? Such as a networking issue.

All errors during `backend` invocation is handled by `.context(CatalogSnafu)` on line 251.

**Review by killme2008 — APPROVED** [review]
LGTM

**Pull request merged**